File: server/src/models/Application.ts
import mongoose, { Document, Schema } from 'mongoose';

export type ApplicationStatus = 'pending' | 'reviewed' | 'shortlisted' | 'rejected';
export type JobSource = 'internal' | 'linkedin' | 'naukri' | 'internshala' | 'unstop';

export interface IExternalJobData {
  title: string;
  company: string;
  location: string;
  applyUrl: string;
  source: JobSource;
  description?: string;
  salary?: string;
}

export interface IApplication extends Document {
  _id: mongoose.Types.ObjectId;
  studentId: mongoose.Types.ObjectId;
  jobId?: mongoose.Types.ObjectId;
  jobSource: JobSource;
  externalJobData?: IExternalJobData;
  coverLetter?: string;
  status: ApplicationStatus;
  notes?: string;
  createdAt: Date;
  updatedAt: Date;
}

const ExternalJobDataSchema = new Schema<IExternalJobData>(
  {
    title: { type: String, required: true },
    company: { type: String, required: true },
    location: { type: String, required: true },
    applyUrl: { type: String, required: true },
    source: { type: String, required: true },
    description: { type: String },
    salary: { type: String },
  },
  { _id: false }
);

const ApplicationSchema = new Schema<IApplication>(
  {
    studentId: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },
    jobId: {
      type: Schema.Types.ObjectId,
      ref: 'Job',
    },
    jobSource: {
      type: String,
      enum: ['internal', 'linkedin', 'naukri', 'internshala', 'unstop'],
      required: true,
    },
    externalJobData: { type: ExternalJobDataSchema },
    coverLetter: {
      type: String,
      maxlength: [2000, 'Cover letter cannot exceed 2000 characters'],
    },
    status: {
      type: String,
      enum: ['pending', 'reviewed', 'shortlisted', 'rejected'],
      default: 'pending',
    },
    notes: { type: String },
  },
  { timestamps: true }
);

// Prevent duplicate applications
ApplicationSchema.index({ studentId: 1, jobId: 1 }, { unique: true, sparse: true });

export default mongoose.model<IApplication>('Application', ApplicationSchema);

File: server/app/models/application.py
from beanie import Document
from pydantic import BaseModel, Field
from typing import Optional, Literal
from datetime import datetime

ApplicationStatus = Literal["pending", "reviewed", "shortlisted", "rejected"]
JobSource = Literal["internal", "linkedin", "naukri", "internshala", "unstop"]


class ExternalJobData(BaseModel):
    title: str
    company: str
    location: str
    apply_url: str = Field(..., alias="applyUrl")
    source: JobSource
    description: Optional[str] = None
    salary: Optional[str] = None

    model_config = {"populate_by_name": True}


class Application(Document):
    student_id: str = Field(..., alias="studentId")
    job_id: Optional[str] = Field(None, alias="jobId")
    job_source: JobSource = Field(..., alias="jobSource")
    external_job_data: Optional[ExternalJobData] = Field(None, alias="externalJobData")
    cover_letter: Optional[str] = Field(None, alias="coverLetter", max_length=2000)
    status: ApplicationStatus = Field(default="pending")
    notes: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "applications"

    model_config = {"populate_by_name": True}

    def to_dict(self) -> dict:
        data = self.model_dump(by_alias=True)
        data["_id"] = str(self.id)
        return data
